File: sections/11_intervals_sweep_line/problems/b_covered_length/solution.py
"""Reference solution for B. Covered Length."""

import sys


def solve_case(intervals: list[tuple[int, int]]) -> int:
    events: list[tuple[int, int]] = []
    for l, r in intervals:
        events.append((l, 1))
        events.append((r, -1))
    events.sort()

    active = 0
    prev = events[0][0]
    total = 0
    i = 0
    while i < len(events):
        x = events[i][0]
        if active > 0:
            total += x - prev
        while i < len(events) and events[i][0] == x:
            active += events[i][1]
            i += 1
        prev = x
    return total


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    t = data[0]
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = data[idx]
        idx += 1
        intervals: list[tuple[int, int]] = []
        for _ in range(n):
            l = data[idx]
            r = data[idx + 1]
            idx += 2
            intervals.append((l, r))
        out.append(str(solve_case(intervals)))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/11_intervals_sweep_line/problems/b_covered_length/solution.cpp
#include <bits/stdc++.h>
using namespace std;

long long solve_case(const vector<pair<long long, long long>> &intervals) {
    vector<pair<long long, int>> events;
    events.reserve(intervals.size() * 2);
    for (auto [l, r] : intervals) {
        events.push_back({l, 1});
        events.push_back({r, -1});
    }
    sort(events.begin(), events.end());

    long long active = 0;
    long long prev = events[0].first;
    long long total = 0;
    int i = 0;
    while (i < (int)events.size()) {
        long long x = events[i].first;
        if (active > 0) {
            total += x - prev;
        }
        while (i < (int)events.size() && events[i].first == x) {
            active += events[i].second;
            ++i;
        }
        prev = x;
    }
    return total;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n;
        cin >> n;
        vector<pair<long long, long long>> intervals(n);
        for (auto &[l, r] : intervals) cin >> l >> r;
        cout << solve_case(intervals) << '\n';
    }

    return 0;
}

